src/utils.ts

```diff
@@ -0,0 +1,275 @@
+    }
+    this.errorStack.push(tmp);
+
+  }
+}
+
+
+/*
+  Copy the contents of one object to another, recursively
+
+  http://stackoverflow.com/questions/12317003/something-like-jquery-extend-but-standalone
+*/
+export
+var extend = function(target: any, source: any): any {
+  target = target || {};
+  for (var prop in source) {
+    if (typeof source[prop] === 'object') {
+      target[prop] = extend(target[prop], source[prop]);
+    } else {
+      target[prop] = source[prop];
+    }
+  }
+  return target;
+};
+
+
+export
+var uuid = function(): string {
+  /**
+   * http://www.ietf.org/rfc/rfc4122.txt
+   */
+  var s: string[] = [];
+  var hexDigits = "0123456789ABCDEF";
+  for (var i = 0; i < 32; i++) {
+    s[i] = hexDigits.substr(Math.floor(Math.random() * 0x10), 1);
+  }
+  s[12] = "4";  // bits 12-15 of the time_hi_and_version field to 0010
+  s[16] = hexDigits.substr((Number(s[16]) & 0x3) | 0x8, 1);  // bits 6-7 of the clock_seq_hi_and_reserved to 01
+
+  var uuid = s.join("");
+  return uuid;
+};
+
+
+/**
+ * join a sequence of url components with '/'
+ */
+export
+var urlPathJoin = function(...paths: string[]): string {
+
+  var url = '';
+  for (var i = 0; i < paths.length; i++) {
+    if (paths[i] === '') {
+      continue;
+    }
+    if (url.length > 0 && url[url.length - 1] != '/') {
+      url = url + '/' + paths[i];
+    } else {
+      url = url + paths[i];
+    }
+  }
+  url = url.replace(/\/\/+/, '/');
+  return url;
+};
+
+
+/**
+ * encode just the components of a multi-segment uri,
+ * leaving '/' separators
+ */
+export
+var encodeURIComponents = function(uri: string): string {
+  return uri.split('/').map(encodeURIComponent).join('/');
+};
+
+
+/**
+ * join a sequence of url components with '/',
+ * encoding each component with encodeURIComponent
+ */
+export
+var urlJoinEncode = function(...args: string[]): string {
+  return encodeURIComponents(urlPathJoin.apply(null, args));
+};
+
+
+// http://stackoverflow.com/questions/2400935/browser-detection-in-javascript
+export
+var browser: string[] = (function() {
+  if (typeof navigator === 'undefined') {
+    // navigator undefined in node
+    return ['None'];
+  }
+  var N: string = navigator.appName;
+  var ua: string = navigator.userAgent
+  var tem: RegExpMatchArray;
+  var M: RegExpMatchArray = ua.match(/(opera|chrome|safari|firefox|msie)\/?\s*(\.?\d+(\.\d+)*)/i);
+  if (M && (tem = ua.match(/version\/([\.\d]+)/i)) !== null) M[2] = tem[1];
+  M = M ? [M[1], M[2]] : [N, navigator.appVersion, '-?'];
+  return M;
+})();
+
+
+/** 
+ * Return a serialized object string suitable for a query
+
+  http://stackoverflow.com/a/30707423
+ */
+export
+var jsonToQueryString = function(json: any) {
+  return '?' +
+    Object.keys(json).map(function(key: string): any {
+      return encodeURIComponent(key) + '=' +
+        encodeURIComponent(json[key]);
+    }).join('&');
+}
+
+
+export
+interface IAjaxSuccess {
+  data: any;
+  status: string;
+  xhr: XMLHttpRequest;
+};
```

Comment:
no semicolon needed at end of class, interface, or function definitions. Only function *expressions* require it.